pkg/util/metric: option to use legacy hdrhistogram model, increase bucket fidelity

dhartunian opened the pull request:
This patch reeintroduces the old HdrHistogram model to optionally be
enabled in favor of the new Prometheus model, gated behind
an environment variable called `COCKROACH_ENABLE_HDR_HISTOGRAMS`,
allowing users a means to "fall back" to the old model in the
event that the new model does not adequately serve their needs
(think of this as an "insurance policy" to protect against
this from happening again with no real mitigation - ideally,
this environment variable should never have to be used).

It also updates the pre-defined bucket boundaries used by the Prometheus
backed histograms with more buckets. This aims to improve precision,
especially for latency histograms, when calculating quantiles (the low precision
being the core cause of the issue at hand).

Note: some histograms were introduced *after* the new
Prometheus histograms were added to CockroachDB. In this
case, we use the `ForceUsePrometheus` option in the
`HistogramOptions` struct to ignore the value of the env
var, since there never was a time where these specific
histograms used the HdrHistogram model.

Release note (ops change): Histogram metrics can now optionally
use the legacy HdrHistogram model by setting the environment var
`COCKROACH_ENABLE_HDR_HISTOGRAMS=true` on CockroachDB nodes.
**Note that this is not recommended** unless users are having
difficulties with the newer Prometheus-backed histogram model.
Enabling can cause performance issues with timeseries databases
like Prometheus, as processing and storing the increased number
of buckets is taxing on both CPU and storage. Note that the
HdrHistogram model is slated for full deprecation in upcoming
releases.

Fixes: https://github.com/cockroachdb/cockroach/issues/95833

dhartunian committed:
pkg/util/metric: optionally reintroduce legacy hdrhistogram model

Addresses https://github.com/cockroachdb/cockroach/issues/95833

This patch reeintroduces the old HdrHistogram model to optionally be
enabled in favor of the new Prometheus model, gated behind
an environment variable called `COCKROACH_ENABLE_HDR_HISTOGRAMS`,
allowing users a means to "fall back" to the old model in the
event that the new model does not adequately serve their needs
(think of this as an "insurance policy" to protect against
this from happening again with no real mitigation - ideally,
this environment variable should never have to be used).

Note: some histograms were introduced *after* the new
Prometheus histograms were added to CockroachDB. In this
case, we use the `ForceUsePrometheus` option in the
`HistogramOptions` struct to ignore the value of the env
var, since there never was a time where these specific
histograms used the HdrHistogram model.

Release note (ops change): Histogram metrics can now optionally
use the legacy HdrHistogram model by setting the environment var
`COCKROACH_ENABLE_HDR_HISTOGRAMS=true` on CockroachDB nodes.
**Note that this is not recommended** unless users are having
difficulties with the newer Prometheus-backed histogram model.
Enabling can cause performance issues with timeseries databases
like Prometheus, as processing and storing the increased number
of buckets is taxing on both CPU and storage. Note that the
HdrHistogram model is slated for full deprecation in upcoming
releases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96029)
<!-- Reviewable:end -->

abarganier committed:
pkg/util/metric: increase bucket counts for Prometheus histograms

This patch increases the fidelity of the histogram buckets for
the new Prometheus model. This is primarily done by increasing the
bucket counts for all latency buckets, but may also be manually
tweaked according to feedback from various engineering teams for
their own use cases.

Release note (ops change): Prometheus histograms will now export
more buckets across the board to improve precision & fidelity of
information reported by histogram metrics, such as quantiles.
This will lead to an increase in storage requirements to process
these histogram metrics in downstream systems like Prometheus,
but should still be a marked improvement when compared to the
legacy HdrHistogram model. If users have issues with the precision
of these bucket boundaries, they can set the environment variable
`COCKROACH_ENABLE_HDR_HISTOGRAMS=true` to revert to using the
legacy HdrHistogram model instead, although this is not recommended
otherwise as the HdrHistogram strains systems like Prometheus with
excessive numbers of histogram buckets. Note that HdrHistograms are
slated for full deprecation in upcoming releases.

abarganier commented:
**REVIEWER TIP:** Start with the changes to files in `pkg/util/metric`. This is where the bulk of the important changes live. The majority of changes in other files are simply migrating existing code to make use of the updated `NewHistogram` constructor function.

miretskiy commented on pkg/util/metric/metric.go:
```diff
@@ -176,10 +185,250 @@ func maybeTick(m periodic) {
+// databases like Prometheus.
+const useHdrHistogramsEnvVar = "COCKROACH_ENABLE_HDR_HISTOGRAMS"
+
+var hdrEnabled = envutil.EnvOrDefaultBool(useHdrHistogramsEnvVar, false)
```
since you're bringing this stuff back, consider making this setting metamorphic:
```
var hdrEnabled = util.ConstantWithMetamorphicTestBool(useHdrHistogramsEnvVar, envutil.EnvOrDefaultBool(....))
```

abarganier replied:
Thanks - I wasn't aware of this function. I changed it over to make use of it.

miretskiy reviewed (commented):
changefeed changes okay; I'm not commenting on the actual metrics stuff.

aadityasondhi reviewed (commented):
Implementation LGTM.

Maybe I lack context here, but just curious about why we are doing both of (1) increasing bucket counts and (2) re-introducing hdr histograms which we worked to remove earlier. In theory, increasing the bucket counts for the affected SQL latencies should give us back the precision lost by the reduction that came with the changes in https://github.com/cockroachdb/cockroach/pull/85990. I think we can resolve the issue of lack of precision by modifying just the prometheus buckets and getting it right. By bringing back hdr, we create additional confusion in the codebase with two different histogram implementations. Using this env var will now also make it so that users running clusters configured differently could have vastly different metric outputs, which could add complexity in debugging.

> Reviewed 36 of 39 files at r1, 2 of 5 files at r3, 5 of 7 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96029#-:-NN82Y5_1B-Nnp2wZwxZ:b-ms7fy3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @jordanlewis, @miretskiy, @tbg, and @ZhouXing19)



<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> Reviewed 39 of 39 files at r1, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96029)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @jordanlewis, @miretskiy, @tbg, and @ZhouXing19)

___
*[`-- commits` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96029#-NN7yWU_407TUePJuDJP):*
Add links to github issue + epic.

The commit message repeats a lot of storytelling from the previous one, which is redundant. I'd prefer if a linked issue did most of the storytelling that these commits referenced.
___
*[`pkg/util/metric/metric_test.go` line 249 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96029#-NN7y7Na4hfgBCLGBQMZ:-NN7y7Na4hfgBCLGBQM_:bg3e831) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f0ce9a38db896c79690aaadd9de05b72fc00aa/pkg/util/metric/metric_test.go#L249)):*
> ```Go
> 		// Windowed histogram is initially empty.
> 		h.Inspect(func(interface{}) {}) // triggers ticking
> 		require.Zero(t, h.TotalSumWindowed())
> ```

is there value to preserving the `Write()` tests to the prometheusgo.Metric object? or is that unused in the codebase?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> If users wish to continue using the
older HdrHistogram model

But this will be phased out, right? We should point that out in the release notes, the idea is really for people to use this as a band-aid, not as a permanent solution, right?

tbg commented on pkg/util/metric/metric.go:
```diff
@@ -17,18 +17,28 @@ import (
 )
 
 const (
+	// MaxLatency is the maximum value tracked in latency histograms. Higher
```
This only applies to `hdrhistograms` - do we want to separate the hdrhistogram code from the current code by moving it into a separate file?

I'm going to assume all constants like these were brought back via a kind of (manual or real) revert.

abarganier replied:
This was done via a manual revert/rework of https://github.com/cockroachdb/cockroach/commit/a82aa8267b1b33a685a02a0f73ec2ece1f8ba295

I've moved all the hdrhistogram code into a separate file as suggested.

tbg commented on pkg/util/metric/metric.go:
```diff
@@ -176,10 +186,261 @@ func maybeTick(m periodic) {
+//
+// Top-level methods generally apply to the cumulative buckets; the windowed
+// variant is exposed through the Windowed method.
+type HdrHistogram struct {
```
Going to assume this is unchanged from before it was removed.

abarganier replied:
Almost! One new function was defined on the `HdrHistogram` to make it adhere to the `IHistogram` interface - `ToPrometheusMetricWindowed()`.

tbg commented on pkg/util/metric/metric.go:
```diff
@@ -176,10 +186,261 @@ func maybeTick(m periodic) {
+	SigFigs            int
+	Buckets            []float64
+	ForceUsePrometheus bool
+	UseHdrLatency      bool
```
Field comments would be helpful, as some of the options are mutually exclusive.

I think you'd really want an enum here,

```go
type HistogramMode byte

const (
  HistogramModePrometheus HistogramMode = iota // corresponds to ForceUsePrometheus
  HistogramModePreferHdr // !ForceUsePrometheus && !UseHdrLatency && hdrEnabled
  HistogramModePreferHdrLatency // ...
)
``` 

but I'm fine w/ cutting that corner.

abarganier replied:
Added field comments, thanks for the reminder!

I also went ahead and added the enum, which I think is a great idea. However, I didn't introduce `HistogramModePreferHdr` as this is communicated via the env var, and I don't think we should ever *prefer* the HDR histograms otherwise. 

Finally, I started at `iota + 1` to avoid mistaking the zero-value with an intentionally set `HistogramMode`.

tbg commented on pkg/util/metric/metric.go:
```diff
@@ -176,10 +186,261 @@ func maybeTick(m periodic) {
+	var sum float64
+	for _, bar := range bars {
+		if bar.Count == 0 {
+			// No need to expose trivial buckets.
```
Alert, this brings a bug back. We do absolutely need to expose trivial buckets. Where did you get this code from? Are there any other changes we're accidentally reverting?

https://github.com/cockroachdb/cockroach/pull/88331/files#diff-7d9143ffc52f8bb21b9a9f1dfdcef120129a14453e54e126e9ba6e3da4b54941R297-R298

tbg replied:
Ah, truly between a rock and a hard place here. Got it, let's use the broken histograms then..

abarganier replied:
Ack - thanks for chiming in @dhartunian. Leaving this as-is, then.

tbg reviewed (changes requested):
Only looked at `metrics.go` so far.

tbg reviewed (approved):
My only blocking comment got resolved - we're intentionally reverting to the broken hdrhistogram because it exposes fewer buckets.

abarganier commented:
> But this will be phased out, right? We should point that out in the release notes, the idea is really for people to use this as a band-aid, not as a permanent solution, right?

Correct! Updated the release notes accordingly - thanks.

abarganier reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96029)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @dhartunian, @irfansharif, @jordanlewis, @miretskiy, @tbg, and @ZhouXing19)

___
*[`-- commits` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96029#-NN7yWU_407TUePJuDJP):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Add links to github issue + epic.

The commit message repeats a lot of storytelling from the previous one, which is redundant. I'd prefer if a linked issue did most of the storytelling that these commits referenced.

</blockquote></details>

Done. LMK if you think it needs more revision.
___
*[`pkg/util/metric/metric_test.go` line 249 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96029#-NN7y7Na4hfgBCLGBQMZ:-NNCu02P8d6ly2PiSIYX:bjcbrwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f0ce9a38db896c79690aaadd9de05b72fc00aa/pkg/util/metric/metric_test.go#L249)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

is there value to preserving the `Write()` tests to the prometheusgo.Metric object? or is that unused in the codebase?
</blockquote></details>

`Write()` would transform the windowed histogram into the `prometheusgo.Metric`, which we then made assertions against.

Now with the new interface we have functions like `TotalSumWindowed()` which [implicitly transformw the hdrhistogram into a windowed Prometheus histogram](https://github.com/cockroachdb/cockroach/pull/96029/commits/5ed760f0f587e8a00a94bbaef2e40334cd63fea2#diff-2dc86a0cbb0c7cd93e6ebf2f5ff832c12a6f4f7055604ca8be11b05c2d7f10acR177-R179) and then pulls values from that, so I don't believe we need this `Write` function anymore since that transformation is handled elsewhere.

cc @aadityasondhi for a sanity check on this.


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96029)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @irfansharif, @jordanlewis, @miretskiy, @tbg, and @ZhouXing19)

___
*[`pkg/util/metric/metric_test.go` line 249 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96029#-NN7y7Na4hfgBCLGBQMZ:-NNDYkZF03rvtPsMFE4p:bma9mjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/27f0ce9a38db896c79690aaadd9de05b72fc00aa/pkg/util/metric/metric_test.go#L249)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

`Write()` would transform the windowed histogram into the `prometheusgo.Metric`, which we then made assertions against.

Now with the new interface we have functions like `TotalSumWindowed()` which [implicitly transformw the hdrhistogram into a windowed Prometheus histogram](https://github.com/cockroachdb/cockroach/pull/96029/commits/5ed760f0f587e8a00a94bbaef2e40334cd63fea2#diff-2dc86a0cbb0c7cd93e6ebf2f5ff832c12a6f4f7055604ca8be11b05c2d7f10acR177-R179) and then pulls values from that, so I don't believe we need this `Write` function anymore since that transformation is handled elsewhere.

cc @aadityasondhi for a sanity check on this.
</blockquote></details>

That sounds correct to me.


<!-- Sent from Reviewable.io -->

abarganier commented:
I've finished testing these changes in roachprod. I wanted to run [these reproduction steps](https://gist.github.com/ZhouXing19/360eab6168ec15b42896a4c72835a870#file-prepare-sh-L27-L32) against the following environments: 
- A build using this commit, with `COCKROACH_ENABLE_HDR_HISTOGRAMS=true`
- A build using this commit, with `COCKROACH_ENABLE_HDR_HISTOGRAMS=false`
- A build using CockroachDB v22.2.2, where the [issue was originally reported](https://github.com/cockroachdb/cockroach/issues/95627).

The results look good. The improvement in precision is very obvious when comparing the v22.2.2 build against this commit's build making use of the new Prometheus histogram buckets. See screenshots from DB Console below.

Based on this test, the core issue seems to be fixed. HDR vs. Prometheus reported latencies are quite similar. When compared to v22.2.2, which uses the old buckets, the improvement is obvious. On the commit builds, reported latency quantiles are nearly equivalent (P90 of ~230ms) between HDR and Prometheus histograms. On the broken v22.2.2 build, reported latency is broken @ ~490ms, which is using the old Prometheus histogram buckets.

There's a bit of a noticeable difference in fidelity still between Prometheus & HDR. Prometheus seems to have a "smoothing" effect on the histogram charts, whereas HDR appears to more accurately capture "peaks" in latency. However, keep in mind that quantile calculations for these HDR histograms are somewhat broken, since empty histogram buckets are omitted for HDR: https://github.com/cockroachdb/cockroach/pull/89532

cc @irfansharif @tbg - curious to hear your opinions on whether these results are acceptable, or if you have any further tests you'd like me to run. Thanks for your continued input 🥇 

### 1. Commit build, `COCKROACH_ENABLE_HDR_HISTOGRAMS=false`

![master-hdr-disabled](https://user-images.githubusercontent.com/8194877/216376023-f34fa8f3-e4f8-4011-b07f-0928a58a4bf0.png)

### 2. v22.2.2 build, using old Prometheus histogram bucket boundaries

![v22 2 2-old-buckets](https://user-images.githubusercontent.com/8194877/216376119-8c6ba2ae-1436-4bf5-893a-7d1b2c006949.png)

### 3. Commit build, `COCKROACH_ENABLE_HDR_HISTOGRAMS=true`

![master-hdr-enabled](https://user-images.githubusercontent.com/8194877/216375940-f7d1413c-76f9-4e05-b460-b89298673644.png)

abarganier commented:
bors r=tbg,aadityasondhi

abarganier commented:
bors r=tbg,aadityasondhi
